### PR TITLE
Index GCS files into elasticsearch

### PR DESCRIPTION
**Pull request by bfcrampton**
@melissachang This isn't quite the format we discussed, though I think it is useful to have the counts to use as a facet in the UI. Can also flatten it to `bam_paths`, bams_count`, etc. if you think that'd be better.

### CODE REVIEW

File: indexer_util/indexer_util/indexer_util.py [hunk 1]
```diff
@@ -101,3 +102,22 @@ def maybe_create_elasticsearch_index(elasticsearch_url, index_name):
             'Creating %s index at %s.' % (index_name, elasticsearch_url))
         es.indices.create(index=index_name, body={})
     return es
+
+
+def bulk_index(es, index, docs_by_id):
```
Comment (melissachang) [inline indexer_util/indexer_util/indexer_util.py]: nit `index_name` to be consistent

File: gcs/indexer.py [hunk 2]
```diff
@@ -37,37 +38,58 @@ def parse_args():
...
-            primary_key = match.group(1)
-            print(
-                'Identified primary key %s from %s' % (primary_key, obj_path))
-        else:
-            raise ValueError('Could not find primary key in %s' % obj_path)
+        path = 'gs://%s/%s' % (bucket_str, obj.name)
+        match = re.match(regex, path)
+        if not match:
+            continue
+        key = match.group(1)
```
Comment (melissachang) [inline gcs/indexer.py]: nit key -> primary_key

File: indexer_util/indexer_util/indexer_util.py [hunk 3]
```diff
@@ -101,3 +102,22 @@ def maybe_create_elasticsearch_index(elasticsearch_url, index_name):
...
+    # Use generator so we can index arbitrarily large iterators (like tables),
+    # without having to load into memory.
+    def es_actions(docs_by_id):
+        for key, doc in docs_by_id:
```
Comment (melissachang) [inline indexer_util/indexer_util/indexer_util.py]: nit key -> id

File: bigquery/indexer.py [hunk 4]
```diff
@@ -5,7 +5,6 @@
 import os
```
Comment (melissachang) [inline bigquery/indexer.py]: We do e2e sanity checking [here](https://github.com/DataBiosphere/data-explorer-indexers/blob/00a8aedd8a89e41d32f4b966f71ee0d9ec0a0ec2/.circleci/config.yml#L18).

Can you add a "Run the GCS indexer" section? Let's just assert that the index has 2559 documents. (You're welcome to add more extensive tests if you want.) Let's also delete the index "curl -XDELETE localhost:9200/1000_genomes" at the end of BQ/GCS sections.

File: gcs/indexer.py [hunk 5]
```diff
@@ -37,37 +38,58 @@ def parse_args():
...
+            if t in path:
+                file_type = t
+        if not file_type:
+            continue
+        _append_to_file_sets(file_sets, key, file_type, path)
+
+    indexer_util.bulk_index(es, index_name, file_sets.iteritems())
+
+
+def _append_to_file_sets(file_sets, key, file_type, path):
```
Comment (melissachang) [inline gcs/indexer.py]: Nit in these repos, we use the convention that helper methods are declared before they're used. Can you move `_append_to_file_sets` to before `_index_gcs_files`?

File: gcs/indexer.py (hunk 2 above)
Comment (bfcrampton) [reply gcs/indexer.py]: Done.

File: gcs/indexer.py (hunk 5 above)
Comment (bfcrampton) [reply gcs/indexer.py]: Done.

File: indexer_util/indexer_util/indexer_util.py (hunk 1 above)
Comment (bfcrampton) [reply indexer_util/indexer_util/indexer_util.py]: Done.

File: indexer_util/indexer_util/indexer_util.py (hunk 3 above)
Comment (bfcrampton) [reply indexer_util/indexer_util/indexer_util.py]: Done (changed to `_id`, `id` is reserved in python)

File: gcs/indexer.py [hunk 6]
```diff
@@ -20,6 +20,7 @@
 logger = logging.getLogger('indexer.gcs')
```
[the remark this answers is not in the document]
Comment (bfcrampton) [reply gcs/indexer.py]: Gotcha okay, removed.

[the remark this answers is not in the document]
Comment (bfcrampton) [reply gcs/indexer.py]: Yeah I think I added that so I could display it as a facet for testing but you're right its duplication, removed.

File: bigquery/indexer.py (hunk 4 above)
Comment (bfcrampton) [reply bigquery/indexer.py]: Ah sorry didn't see that CI testing, added GCS indexing section.

File: .circleci/config.yml [hunk 7]
```diff
@@ -27,6 +27,22 @@ jobs:
...
+          command: |
+            echo ${GOOGLE_SERVICE_KEY} | base64 --decode > ${HOME}/.config/gcloud/application_default_credentials.json
+            # Note: data-explorer_default has already been created
+            cd gcs
+            # Run Elasticsearch in the background with the indexer in the foreground to prevent blocking the main thread
+            docker-compose up -d elasticsearch
+            BILLING_PROJECT_ID=${GOOGLE_PROJECT_ID} docker-compose up --build indexer
+            # For some reason index isn't available right after indexer terminates, so sleep
+            sleep 5
+            # Assert the 1000_genomes index has 3500 documents
```
Comment (melissachang) [inline .circleci/config.yml]: Comment out of date. Delete comment and line 28. I figure it's easier to delete than keep up to date...

File: .circleci/config.yml [hunk 8]
```diff
@@ -27,6 +27,22 @@ jobs:
...
+            expr $(curl -s 'http://localhost:9200/1000_genomes/_search' | jq -r '.hits.total') = "2559"
+            curl -XDELETE localhost:9200/1000_genomes
```
Comment (melissachang) [inline .circleci/config.yml]: Can you add delete for BigQuery indexer also?

Comment (bfcrampton) [reply .circleci/config.yml]: https://github.com/DataBiosphere/data-explorer-indexers/pull/48/files/589083b6440e5addd28c6d90938d09127ec56840#diff-1d37e48f9ceff6d8030570cd36286a61R30

File: .circleci/config.yml (hunk 7 above)
Comment (bfcrampton) [reply .circleci/config.yml]: Done.

**Pull request merged**